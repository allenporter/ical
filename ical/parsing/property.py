"""Library for handling rfc5545 properties and parameters.

A property is the definition of an individual attribute describing a
calendar object or a calendar component. A property is also really
just a "contentline", however properties in this file are the
output of the parser and are provided in the context of where
they live on a component hierarchy (e.g. attached to a component,
or sub component).
"""

# mypy: allow-any-generics

from __future__ import annotations

import datetime
from dataclasses import dataclass
from typing import Optional, Union

from .const import PARSE_PARAM_NAME, PARSE_PARAM_VALUE, PARSE_PARAMS
from .unicode import UNSAFE_CHAR_RE


@dataclass
class ParsedPropertyParameter:
    """An rfc5545 property parameter."""

    name: str

    values: list[Union[str, datetime.tzinfo]]
    """Values are typically strings, with a hack for TZID.

    The values may be overridden in the parse tree so that we can directly
    set the timezone information when parsing a date-time rather than
    combining with the calendar at runtime. That is, we update the tree
    with timezone infrmation replacing a string TZID with the zoneinfo.
    """


@dataclass
class ParsedProperty:
    """An rfc5545 property."""

    name: str
    value: str
    params: Optional[list[ParsedPropertyParameter]] = None

    def get_parameter(self, name: str) -> ParsedPropertyParameter | None:
        """Return a single ParsedPropertyParameter with the specified name."""
        if not self.params:
            return None
        for param in self.params:
            if param.name.lower() != name.lower():
                continue
            return param
        return None

    def get_parameter_value(self, name: str) -> str | None:
        """Return the property parameter value."""
        if not (param := self.get_parameter(name)):
            return None
        if len(param.values) > 1:
            raise ValueError(
                f"Expected only a single parameter string value, got {param.values}"
            )
        return param.values[0] if isinstance(param.values[0], str) else None

    def ics(self) -> str:
        """Encode a ParsedProperty into the serialized format."""
        result = [self.name.upper()]
        if self.params:
            result.append(";")
            result_params = []
            for parameter in self.params:
                result_param_values = []
                for value in parameter.values:
                    if not isinstance(value, str):
                        continue  # Shouldn't happen; only strings are set by parsing
                    # Property parameters with values contain a colon, simicolon,
                    # or a comma character must be placed in quoted text
                    if UNSAFE_CHAR_RE.search(value):
                        result_param_values.append(f'"{value}"')
                    else:
                        result_param_values.append(value)
                values = ",".join(result_param_values)
                result_params.append(f"{parameter.name.upper()}={values}")
            result.append(";".join(result_params))
        result.append(":")
        result.append(str(self.value))
        return "".join(result)


def parse_property_params(
    parse_result_dict: dict[str, str | list]
) -> list[ParsedPropertyParameter]:
    """Extract the property parameters from a pyparsing ParseResult object."""
    if PARSE_PARAMS not in parse_result_dict:
        return []
    params: list[ParsedPropertyParameter] = []
    for parsed_params in parse_result_dict[PARSE_PARAMS]:
        if not isinstance(parsed_params, dict) or PARSE_PARAMS not in parsed_params:
            continue
        for pair in parsed_params[PARSE_PARAMS]:
            params.append(
                ParsedPropertyParameter(
                    name=pair[PARSE_PARAM_NAME], values=pair[PARSE_PARAM_VALUE]
                )
            )
    return params
