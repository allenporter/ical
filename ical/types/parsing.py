"""Library for parsing rfc5545 types."""

from __future__ import annotations

import logging
from typing import Any

try:
    from pydantic.v1 import BaseModel
    from pydantic.v1.fields import SHAPE_LIST, ModelField
except ImportError:
    from pydantic import BaseModel
    from pydantic.fields import SHAPE_LIST, ModelField

_LOGGER = logging.getLogger(__name__)


def _all_fields(cls: BaseModel) -> dict[str, ModelField]:
    all_fields: dict[str, ModelField] = {}
    for model_field in cls.__fields__.values():
        all_fields[model_field.name] = model_field
        all_fields[model_field.alias] = model_field
    return all_fields


def parse_parameter_values(cls: BaseModel, values: dict[str, Any]) -> dict[str, Any]:
    """Convert property parameters to pydantic fields."""
    _LOGGER.debug("parse_parameter_values=%s", values)
    if params := values.get("params"):
        all_fields = _all_fields(cls)
        for param in params:
            if not (field := all_fields.get(param["name"])):
                continue
            if field.shape == SHAPE_LIST:
                values[param["name"]] = param["values"]
            else:
                if len(param["values"]) > 1:
                    raise ValueError("Unexpected repeated property parameter")
                values[param["name"]] = param["values"][0]
    return values
