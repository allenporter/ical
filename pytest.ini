[pytest]
log_level = DEBUG
# Can be run manually with --benchmark-only
addopts = --benchmark-skip
