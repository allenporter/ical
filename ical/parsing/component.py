"""Library for handling rfc5545 components.

An iCalendar object consists of one or more components, that may have
properties or sub-components. An example of a component might be the
calendar itself, an event, a to-do, a journal entry, timezone info, etc.

Components created here have no semantic meaning, but hold all the
data needed to interpret based on the type (e.g. by a pydantic model)
"""

# mypy: allow-any-generics

from __future__ import annotations

import re
import textwrap
from dataclasses import dataclass, field

from .const import (
    ATTR_BEGIN,
    ATTR_END,
    FOLD,
    FOLD_INDENT,
    FOLD_LEN,
    PARSE_NAME,
    PARSE_PARAMS,
    PARSE_VALUE,
)
from .parser import parse_contentlines
from .property import ParsedProperty, parse_property_params


@dataclass
class ParsedComponent:
    """An rfc5545 component."""

    name: str
    properties: list[ParsedProperty] = field(default_factory=list)
    components: list[ParsedComponent] = field(default_factory=list)

    def as_dict(self) -> dict[str, str | list[ParsedProperty | dict]]:
        """Convert the component into a pydantic parseable dictionary."""
        result: dict[str, list[ParsedProperty | dict]] = {}
        for prop in self.properties:
            result.setdefault(prop.name, [])
            result[prop.name].append(prop)
        for component in self.components:
            result.setdefault(component.name, [])
            result[component.name].append(component.as_dict())
        return {
            "name": self.name,
            **result,
        }

    def ics(self) -> str:
        """Encode a component as rfc5545 text."""
        contentlines = []
        name = self.name.upper()
        contentlines.append(f"{ATTR_BEGIN}:{name}")
        for prop in self.properties:
            contentlines.extend(_fold(prop.ics()))
        contentlines.extend([component.ics() for component in self.components])
        contentlines.append(f"{ATTR_END}:{name}")
        return "\n".join(contentlines)


def _fold(contentline: str) -> list[str]:
    return textwrap.wrap(
        contentline,
        width=FOLD_LEN,
        subsequent_indent=FOLD_INDENT,
        drop_whitespace=False,
        replace_whitespace=False,
        expand_tabs=False,
        break_on_hyphens=False,
    )


def parse_content(content: str) -> list[ParsedComponent]:
    """Parse content into raw properties.

    This includes all necessary unfolding of long lines into full properties.

    This is fairly straight forward in that it walks through each line and uses
    a stack to associate properties with the current object. This does the absolute
    minimum possible parsing into a dictionary of objects to get the right structure.
    All the more detailed parsing of the objects is handled by pydantic, elsewhere.
    """
    content = re.sub(FOLD, "", content, flags=re.MULTILINE)  # Unfold content lines
    lines = re.split("\r?\n", content)
    token_results = parse_contentlines(lines)

    stack: list[ParsedComponent] = [ParsedComponent(name="stream")]
    for result in token_results:
        result_dict = result.as_dict()
        if PARSE_NAME not in result_dict:
            raise ValueError(
                f"Missing fields {PARSE_NAME} or {PARSE_VALUE} in {result_dict}"
            )
        name = result_dict[PARSE_NAME]
        value = result_dict.get(PARSE_VALUE, "")
        if name == ATTR_BEGIN:
            stack.append(ParsedComponent(name=value.lower()))
        elif name == ATTR_END:
            value = value.lower()
            component = stack.pop()
            if value != component.name:
                raise ValueError(
                    f"Unexpected '{result}', expected {ATTR_END}:{component.name}"
                )
            stack[-1].components.append(component)
        else:
            name = name.lower()
            property_dict = {
                PARSE_NAME: name,
                PARSE_VALUE: value,
            }
            if property_params := parse_property_params(result_dict):
                property_dict[PARSE_PARAMS] = property_params
            stack[-1].properties.append(ParsedProperty(**property_dict))
    return stack[0].components


def encode_content(components: list[ParsedComponent]) -> str:
    """Encode a set of parsed properties into content."""
    return "\n".join([component.ics() for component in components])
