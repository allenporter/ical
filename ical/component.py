"""Library for parsing and encoding rfc5545 components with pydantic.

The data model returned by the contentlines parsing is a bag of ParsedProperty
objects that support all the flexibility of the rfc5545 spec. However in the
common case the spec has a lot more flexibility than is needed for handling
simple property types e.g. a single summary field that is specified only once.

This library helps reduce boilerplate for translating that complex structure
into the simpler pydantic data model, and handles custom field types and
validators.

Just as the pydantic model provides syntax glue for parsing data and
associating necessary validators, this is also doing the same thing
in the opposite direction to encode back to ICS.
"""

from __future__ import annotations

import copy
import datetime
import json
import logging
from typing import Any, Union, get_args, get_origin

try:
    from pydantic.v1 import BaseModel, root_validator
    from pydantic.v1.fields import SHAPE_LIST
except ImportError:
    from pydantic import BaseModel, root_validator
    from pydantic.fields import SHAPE_LIST

from .parsing.component import ParsedComponent
from .parsing.property import ParsedProperty
from .types.data_types import DATA_TYPE

_LOGGER = logging.getLogger(__name__)


ATTR_VALUE = "VALUE"

# Repeated values can either be specified as multiple separate values, but
# also some values support repeated values within a single value with a
# comma delimiter, listed here.
EXPAND_REPEATED_VALUES = {
    "categories",
    "classification",
    "exdate",
    "rdate",
    "resources",
    "freebusy",
}


def _adjust_recurrence_date(
    date_value: datetime.datetime | datetime.date,
    dtstart: datetime.datetime | datetime.date,
) -> datetime.datetime | datetime.date:
    """Apply fixes to the recurrence rule date."""
    if isinstance(dtstart, datetime.datetime):
        if not isinstance(date_value, datetime.datetime):
            raise ValueError(
                "DTSTART was DATE-TIME but UNTIL was DATE: "
                "must be the same value type"
            )
        if dtstart.tzinfo is None:
            if date_value.tzinfo is not None:
                raise ValueError("DTSTART is date local but UNTIL was not")
            return date_value

        if date_value.utcoffset():
            raise ValueError("DTSTART had UTC or local and UNTIL must be UTC")

        return date_value

    if isinstance(date_value, datetime.datetime):
        # Fix invalid rules where UNTIL value is DATE-TIME but DTSTART is DATE
        return date_value.date()

    return date_value


def validate_until_dtstart(_cls: BaseModel, values: dict[str, Any]) -> dict[str, Any]:
    """Verify the until time and dtstart are the same."""
    if (
        not (rule := values.get("rrule"))
        or not rule.until
        or not (dtstart := values.get("dtstart"))
    ):
        return values
    rule.until = _adjust_recurrence_date(rule.until, dtstart)
    return values


def _as_datetime(
    date_value: datetime.datetime | datetime.date,
    dtstart: datetime.datetime,
) -> datetime.datetime:
    if not isinstance(date_value, datetime.datetime):
        return datetime.datetime.combine(date_value, dtstart.time())
    return date_value


def validate_recurrence_dates(
    _cls: BaseModel, values: dict[str, Any]
) -> dict[str, Any]:
    """Verify the recurrence dates have the correct types."""
    if (
        not values.get("rrule")
        or not (dtstart := values.get("dtstart"))
        or not isinstance(dtstart, datetime.datetime)
    ):
        return values
    for field in ("exdate", "rdate"):
        if not (date_values := values.get(field)):
            continue
        values[field] = [
            _as_datetime(date_value, dtstart) for date_value in date_values
        ]
    return values


class ComponentModel(BaseModel):
    """Abstract class for rfc5545 component model."""

    @root_validator(pre=True, allow_reuse=True)
    def parse_extra_fields(
        cls, values: dict[str, list[ParsedProperty | ParsedComponent]]
    ) -> dict[str, Any]:
        """Parse extra fields not in the model."""
        all_fields = set()
        for field in cls.__fields__.values():
            all_fields |= {field.alias, field.name}

        extras: list[ParsedProperty | ParsedComponent] = []
        for field_name, value in values.items():
            if field_name in all_fields:
                continue
            for prop in value:
                if isinstance(prop, ParsedProperty):
                    extras.append(prop)
        if extras:
            values["extras"] = extras
        return values

    @root_validator(pre=True, allow_reuse=True)
    def parse_property_values(cls, values: dict[str, Any]) -> dict[str, Any]:
        """Parse individual ParsedProperty value fields."""
        _LOGGER.debug("Parsing value data %s", values)

        for field in cls.__fields__.values():
            if field.alias == "extras":
                continue
            if not (value := values.get(field.alias)):
                continue
            if not (isinstance(value, list) and isinstance(value[0], ParsedProperty)):
                # The incoming value is not from the parse tree
                continue
            if field.alias in EXPAND_REPEATED_VALUES:
                value = cls._expand_repeated_property(value)
            # Repeated values will accept a list, otherwise truncate to a single
            # value when repeated is not allowed.
            allow_repeated = field.shape == SHAPE_LIST
            if not allow_repeated and len(value) > 1:
                raise ValueError(f"Expected one value for field: {field.alias}")
            field_types = cls._get_field_types(field.type_)
            validated = [cls._parse_property(field_types, prop) for prop in value]
            values[field.alias] = validated if allow_repeated else validated[0]

        _LOGGER.debug("Completed parsing value data %s", values)

        return values

    @classmethod
    def _parse_property(cls, field_types: list[type], prop: ParsedProperty) -> Any:
        """Parse an individual field value from a ParsedProperty as the specified types."""
        _LOGGER.debug(
            "Parsing field '%s' with value '%s' as types %s",
            prop.name,
            prop.value,
            field_types,
        )
        errors = []
        for sub_type in field_types:
            try:
                return cls._parse_single_property(sub_type, prop)
            except ValueError as err:
                _LOGGER.debug(
                    "Unable to parse property value as type %s: %s", sub_type, err
                )
                errors.append(str(err))
                continue
        raise ValueError(f"Failed to validate: {prop.value}, errors: ({errors})")

    @classmethod
    def _parse_single_property(cls, field_type: type, prop: ParsedProperty) -> Any:
        """Parse an individual field as a single type."""
        if (
            value_type := prop.get_parameter_value(ATTR_VALUE)
        ) and field_type not in DATA_TYPE.disable_value_param:
            # Property parameter specified a strong type
            if func := DATA_TYPE.parse_parameter_by_name.get(value_type):
                _LOGGER.debug("Parsing %s as value type '%s'", prop.name, value_type)
                return func(prop)
            # Consider graceful degradation instead in the future
            raise ValueError(
                f"Property parameter specified unsupported type: {value_type}"
            )

        if decoder := DATA_TYPE.parse_property_value.get(field_type):
            _LOGGER.debug("Decoding '%s' as type '%s'", prop.name, field_type)
            return decoder(prop)

        _LOGGER.debug("Using '%s' bare property value '%s'", prop.name, prop.value)
        return prop.value

    @classmethod
    def _expand_repeated_property(
        cls, value: list[ParsedProperty]
    ) -> list[ParsedProperty]:
        """Expand properties with repeated values into separate properties."""
        result: list[ParsedProperty] = []
        for prop in value:
            if "," in prop.value:
                for sub_value in prop.value.split(","):
                    sub_prop = copy.deepcopy(prop)
                    sub_prop.value = sub_value
                    result.append(sub_prop)
            else:
                result.append(prop)
        return result

    @classmethod
    def _get_field_types(cls, field_type: type) -> list[type]:
        """Return type to attempt for encoding/decoding based on the field type."""
        origin = get_origin(field_type)
        if origin is Union:
            if not (args := get_args(field_type)):
                raise ValueError(f"Unable to determine args of type: {field_type}")

            # get_args does not have a deterministic order, so use the order supplied
            # in the registry. Ignore None as its not a parseable type.
            sortable_args = [
                (DATA_TYPE.parse_order.get(arg, 0), arg)
                for arg in args
                if arg is not type(None)  # noqa: E721
            ]
            sortable_args.sort(reverse=True)
            return [arg for (order, arg) in sortable_args]
        return [field_type]

    def __encode_component_root__(self) -> ParsedComponent:
        """Encode the calendar stream as an rfc5545 iCalendar content."""
        # The overall data model hierarchy is created by pydantic and properties
        # are encoded using the json encoders specific for each type. These are
        # marshalled through as string values. There are then additional passes
        # to ge the data in to the right final format for ics encoding.
        model_data = json.loads(
            self.json(by_alias=True, exclude_none=True, exclude_defaults=True)
        )
        # The component name is ignored as we're really only encoding children components
        return self.__encode_component__(self.__class__.__name__, model_data)

    @classmethod
    def __encode_component__(
        cls, name: str, model_data: dict[str, Any]
    ) -> ParsedComponent:
        """Encode this object as a component to prepare for serialization.

        The data passed in have already been encoded with one pass from the root json
        encoder. This method takes additional passes to add more field specific encoding,
        as well as overall component objects.
        """
        parent = ParsedComponent(name=name)
        for field in cls.__fields__.values():
            key = field.alias
            values = model_data.get(key)
            if values is None or key == "extras":
                continue
            if not isinstance(values, list):
                values = [values]
            for value in values:
                if component_encoder := getattr(
                    field.type_, "__encode_component__", None
                ):
                    parent.components.append(component_encoder(key, value))
                    continue
                if prop := cls._encode_property(key, field.type_, value):
                    parent.properties.append(prop)
        return parent

    @classmethod
    def _encode_property(cls, key: str, field_type: type, value: Any) -> ParsedProperty:
        """Encode an individual property for the specified field."""
        # A property field may have multiple possible types, like for
        # a Union. Pick the first type that is able to encode the value.
        errors = []
        for sub_type in cls._get_field_types(field_type):
            encoded_value: Any | None = None
            if value_encoder := DATA_TYPE.encode_property_value.get(sub_type):
                try:
                    encoded_value = value_encoder(value)
                except ValueError as err:
                    _LOGGER.debug("Encoding failed for property: %s", err)
                    errors.append(str(err))
                    continue
            else:
                encoded_value = value

            if encoded_value is not None:
                prop = ParsedProperty(name=key, value=encoded_value)
                if params_encoder := DATA_TYPE.encode_property_params.get(
                    sub_type, None
                ):
                    if params := params_encoder(value):
                        prop.params = params
                return prop

        raise ValueError(f"Unable to encode property: {value}, errors: {errors}")

    class Config:
        """Pyandtic model configuration."""

        validate_assignment = True
        allow_population_by_field_name = True
        smart_union = True
        arbitrary_types_allowed = True
