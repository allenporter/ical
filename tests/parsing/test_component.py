"""Tests for parsing raw components."""

import json
from typing import Any

import pytest
from pytest_golden.plugin import GoldenTestFixture

from ical.parsing.component import encode_content, parse_content


@pytest.mark.golden_test("testdata/*.yaml")
def test_parse_contentlines(
    golden: GoldenTestFixture, json_encoder: json.JSONEncoder
) -> None:
    """Fixture to read golden file and compare to golden output."""
    values = parse_content(golden["input"])
    values = json.loads(json_encoder.encode(values))
    assert values == golden["output"]


@pytest.mark.golden_test("testdata/*.yaml")
def test_encode_contentlines(golden: GoldenTestFixture) -> None:
    """Fixture to read golden file and serialize back to same format."""
    values = parse_content(golden["input"])
    ics = encode_content(values)
    assert ics == golden.get("encoded", golden["input"])


@pytest.mark.golden_test("testdata/*.yaml")
def test_parse_contentlines_benchmark(
    golden: GoldenTestFixture, json_encoder: json.JSONEncoder, benchmark: Any
) -> None:
    """Benchmark to measure the speed of parsing."""

    def parse() -> None:
        values = parse_content(golden["input"])
        values = json.loads(json_encoder.encode(values))
        assert values == golden["output"]

    benchmark(parse)
