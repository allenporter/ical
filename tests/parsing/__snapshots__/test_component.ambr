# serializer version: 1
# name: test_encode_contentlines[attendee]
  '''
  BEGIN:VEVENT
  ATTENDEE;RSVP=TRUE;ROLE=REQ-PARTICIPANT:mailto: jsmith@example.com
  ATTENDEE;DELEGATED-TO="mailto:jdoe@example.com","mailto:jqpublic 
   @example.com":mailto:jsmith@example.com
  END:VEVENT
  '''
# ---
# name: test_encode_contentlines[comma]
  '''
  BEGIN:VEVENT
  DESCRIPTION;ALTREP="cid:part1.0001@example.org":The Fall'98 Wild Wizards 
   Conference - - Las Vegas\, NV\, USA
  END:VEVENT
  '''
# ---
# name: test_encode_contentlines[fold]
  '''
  BEGIN:VEVENT
  DESCRIPTION:This is a lo ng description that exists on a long line.
  END:VEVENT
  '''
# ---
# name: test_encode_contentlines[icalendar_object]
  '''
  BEGIN:VCALENDAR
  VERSION:2.0
  PRODID:-//hacksw/handcal//NONSGML v1.0//EN
  BEGIN:VEVENT
  UID:19970610T172345Z-AF23B2@example.com
  DTSTAMP:19970610T172345Z
  DTSTART:19970714T170000Z
  DTEND:19970715T040000Z
  SUMMARY:Bastille Day Party
  END:VEVENT
  END:VCALENDAR
  '''
# ---
# name: test_encode_contentlines[params]
  '''
  BEGIN:VCALENDAR
  NAME;PARAM-NAME=PARAM-VALUE1:VALUE
  NAME;PARAM-NAME=PARAM+VALUE2:VALUE
  NAME;PARAM-NAME=PARAM VALUE3:VALUE
  NAME;PARAM-NAME="PARAM:VALUE4":VALUE
  DESCRIPTION:
  DESCRIPTION;TYPE=:
  END:VCALENDAR
  '''
# ---
# name: test_encode_contentlines[params_quoted]
  '''
  BEGIN:VCALENDAR
  NAME;PARAM-NAME=PARAM-VALUE:VALUE
  NAME;PARAM-NAME="PARAM:VALUE":VALUE
  END:VCALENDAR
  '''
# ---
# name: test_encode_contentlines[rdate]
  '''
  BEGIN:VCALENDAR
  RDATE;VALUE=DATE:19970304,19970504,19970704,19970904
  END:VCALENDAR
  '''
# ---
# name: test_encode_contentlines[unicode]
  '''
  BEGIN:VCALENDAR
  PRODID:-//ABC Corporation//NONSGML My Product//EN
  VERSION:2.0
  BEGIN:VTODO
  DTSTAMP:19980130T134500Z
  SEQUENCE:2
  UID:uid4@example.com
  DUE:19980415T000000
  STATUS:NEEDS-ACTION
  SUMMARY:Birthday
  BEGIN:VALARM
  ACTION:AUDIO
  TRIGGER:19980403T120000Z
  ATTACH;FILENAME=FÃ¸dselsdag_40.pdf:https://someurl.com
  REPEAT:4
  DURATION:PT1H
  END:VALARM
  END:VTODO
  END:VCALENDAR
  '''
# ---
# name: test_encode_contentlines[vcalendar_emoji]
  '''
  BEGIN:VEVENT
  DTSTAMP:20221202T075310
  UID:5deea302-7216-11ed-b1b6-48d2240d04ae
  DTSTART:20221202T085500
  DTEND:20221202T090000
  SUMMARY:ðŸŽ„emojis!
  CREATED:20221202T075310
  SEQUENCE:0
  END:VEVENT
  '''
# ---
# name: test_encode_contentlines[vevent]
  '''
  BEGIN:VEVENT
  UID:19970901T130000Z-123401@example.com
  DTSTAMP:19970901T130000Z
  DTSTART:19970903T163000Z
  DTEND:19970903T190000Z
  SUMMARY:Annual Employee Review
  CLASS:PRIVATE
  CATEGORIES:BUSINESS,HUMAN RESOURCES
  END:VEVENT
  '''
# ---
# name: test_encode_contentlines[vtodo]
  '''
  BEGIN:VTODO
  UID:20070313T123432Z-456553@example.com
  DTSTAMP:20070313T123432Z
  DUE;VALUE=DATE:20070501
  SUMMARY:Submit Quebec Income Tax Return for 2006
  CLASS:CONFIDENTIAL
  CATEGORIES:FAMILY,FINANCE
  STATUS:NEEDS-ACTION
  END:VTODO
  '''
# ---
# name: test_encode_contentlines[x-name]
  '''
  BEGIN:VEVENT
  X-DESCRIPTION:This is a description
  X-ICAL-DESCRIPTION:This is a description
  X-ICAL-1:This is another x-name
  END:VEVENT
  '''
# ---
# name: test_parse_contentlines[attendee]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'attendee',
          'params': list([
            dict({
              'name': 'RSVP',
              'values': list([
                'TRUE',
              ]),
            }),
            dict({
              'name': 'ROLE',
              'values': list([
                'REQ-PARTICIPANT',
              ]),
            }),
          ]),
          'value': 'mailto: jsmith@example.com',
        }),
        dict({
          'name': 'attendee',
          'params': list([
            dict({
              'name': 'DELEGATED-TO',
              'values': list([
                'mailto:jdoe@example.com',
                'mailto:jqpublic @example.com',
              ]),
            }),
          ]),
          'value': 'mailto:jsmith@example.com',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[comma]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'description',
          'params': list([
            dict({
              'name': 'ALTREP',
              'values': list([
                'cid:part1.0001@example.org',
              ]),
            }),
          ]),
          'value': "The Fall'98 Wild Wizards Conference - - Las Vegas\\, NV\\, USA",
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[fold]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'description',
          'params': None,
          'value': 'This is a lo ng description that exists on a long line.',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[icalendar_object]
  list([
    dict({
      'components': list([
        dict({
          'components': list([
          ]),
          'name': 'vevent',
          'properties': list([
            dict({
              'name': 'uid',
              'params': None,
              'value': '19970610T172345Z-AF23B2@example.com',
            }),
            dict({
              'name': 'dtstamp',
              'params': None,
              'value': '19970610T172345Z',
            }),
            dict({
              'name': 'dtstart',
              'params': None,
              'value': '19970714T170000Z',
            }),
            dict({
              'name': 'dtend',
              'params': None,
              'value': '19970715T040000Z',
            }),
            dict({
              'name': 'summary',
              'params': None,
              'value': 'Bastille Day Party',
            }),
          ]),
        }),
      ]),
      'name': 'vcalendar',
      'properties': list([
        dict({
          'name': 'version',
          'params': None,
          'value': '2.0',
        }),
        dict({
          'name': 'prodid',
          'params': None,
          'value': '-//hacksw/handcal//NONSGML v1.0//EN',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[params]
  list([
    dict({
      'name': 'vcalendar',
      'properties': list([
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM-VALUE1',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM+VALUE2',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM VALUE3',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM:VALUE4',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
        dict({
          'name': 'description',
          'params': None,
          'value': '',
        }),
        dict({
          'name': 'description',
          'params': list([
            dict({
              'name': 'TYPE',
              'values': list([
                '',
              ]),
            }),
          ]),
          'value': '',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[params_quoted]
  list([
    dict({
      'name': 'vcalendar',
      'properties': list([
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM-VALUE',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
        dict({
          'name': 'name',
          'params': list([
            dict({
              'name': 'PARAM-NAME',
              'values': list([
                'PARAM:VALUE',
              ]),
            }),
          ]),
          'value': 'VALUE',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[rdate]
  list([
    dict({
      'name': 'vcalendar',
      'properties': list([
        dict({
          'name': 'rdate',
          'params': list([
            dict({
              'name': 'VALUE',
              'values': list([
                'DATE',
              ]),
            }),
          ]),
          'value': '19970304,19970504,19970704,19970904',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[unicode]
  list([
    dict({
      'components': list([
        dict({
          'components': list([
            dict({
              'components': list([
              ]),
              'name': 'valarm',
              'properties': list([
                dict({
                  'name': 'action',
                  'params': None,
                  'value': 'AUDIO',
                }),
                dict({
                  'name': 'trigger',
                  'params': None,
                  'value': '19980403T120000Z',
                }),
                dict({
                  'name': 'attach',
                  'params': list([
                    dict({
                      'name': 'FILENAME',
                      'values': list([
                        'FÃ¸dselsdag_40.pdf',
                      ]),
                    }),
                  ]),
                  'value': 'https://someurl.com',
                }),
                dict({
                  'name': 'repeat',
                  'params': None,
                  'value': '4',
                }),
                dict({
                  'name': 'duration',
                  'params': None,
                  'value': 'PT1H',
                }),
              ]),
            }),
          ]),
          'name': 'vtodo',
          'properties': list([
            dict({
              'name': 'dtstamp',
              'params': None,
              'value': '19980130T134500Z',
            }),
            dict({
              'name': 'sequence',
              'params': None,
              'value': '2',
            }),
            dict({
              'name': 'uid',
              'params': None,
              'value': 'uid4@example.com',
            }),
            dict({
              'name': 'due',
              'params': None,
              'value': '19980415T000000',
            }),
            dict({
              'name': 'status',
              'params': None,
              'value': 'NEEDS-ACTION',
            }),
            dict({
              'name': 'summary',
              'params': None,
              'value': 'Birthday',
            }),
          ]),
        }),
      ]),
      'name': 'vcalendar',
      'properties': list([
        dict({
          'name': 'prodid',
          'params': None,
          'value': '-//ABC Corporation//NONSGML My Product//EN',
        }),
        dict({
          'name': 'version',
          'params': None,
          'value': '2.0',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[vcalendar_emoji]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'dtstamp',
          'params': None,
          'value': '20221202T075310',
        }),
        dict({
          'name': 'uid',
          'params': None,
          'value': '5deea302-7216-11ed-b1b6-48d2240d04ae',
        }),
        dict({
          'name': 'dtstart',
          'params': None,
          'value': '20221202T085500',
        }),
        dict({
          'name': 'dtend',
          'params': None,
          'value': '20221202T090000',
        }),
        dict({
          'name': 'summary',
          'params': None,
          'value': 'ðŸŽ„emojis!',
        }),
        dict({
          'name': 'created',
          'params': None,
          'value': '20221202T075310',
        }),
        dict({
          'name': 'sequence',
          'params': None,
          'value': '0',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[vevent]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'uid',
          'params': None,
          'value': '19970901T130000Z-123401@example.com',
        }),
        dict({
          'name': 'dtstamp',
          'params': None,
          'value': '19970901T130000Z',
        }),
        dict({
          'name': 'dtstart',
          'params': None,
          'value': '19970903T163000Z',
        }),
        dict({
          'name': 'dtend',
          'params': None,
          'value': '19970903T190000Z',
        }),
        dict({
          'name': 'summary',
          'params': None,
          'value': 'Annual Employee Review',
        }),
        dict({
          'name': 'class',
          'params': None,
          'value': 'PRIVATE',
        }),
        dict({
          'name': 'categories',
          'params': None,
          'value': 'BUSINESS,HUMAN RESOURCES',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[vtodo]
  list([
    dict({
      'name': 'vtodo',
      'properties': list([
        dict({
          'name': 'uid',
          'params': None,
          'value': '20070313T123432Z-456553@example.com',
        }),
        dict({
          'name': 'dtstamp',
          'params': None,
          'value': '20070313T123432Z',
        }),
        dict({
          'name': 'due',
          'params': list([
            dict({
              'name': 'VALUE',
              'values': list([
                'DATE',
              ]),
            }),
          ]),
          'value': '20070501',
        }),
        dict({
          'name': 'summary',
          'params': None,
          'value': 'Submit Quebec Income Tax Return for 2006',
        }),
        dict({
          'name': 'class',
          'params': None,
          'value': 'CONFIDENTIAL',
        }),
        dict({
          'name': 'categories',
          'params': None,
          'value': 'FAMILY,FINANCE',
        }),
        dict({
          'name': 'status',
          'params': None,
          'value': 'NEEDS-ACTION',
        }),
      ]),
    }),
  ])
# ---
# name: test_parse_contentlines[x-name]
  list([
    dict({
      'name': 'vevent',
      'properties': list([
        dict({
          'name': 'x-description',
          'params': None,
          'value': 'This is a description',
        }),
        dict({
          'name': 'x-ical-description',
          'params': None,
          'value': 'This is a description',
        }),
        dict({
          'name': 'x-ical-1',
          'params': None,
          'value': 'This is another x-name',
        }),
      ]),
    }),
  ])
# ---
